### DISCUSSION TITLE
Where is the mapping between labels and zigbee attribute IDs and where is calculation done for thermostat_outdoor_temperature ?

### DISCUSSION THREAD

**Discussion by mboisson** [Koenkk/zigbee2mqtt#15423] [Devices]
Hi,
I have been trying to update a Hubitat zigbee driver for the Stelpro Allia thermostats, which were recently added in zigbee2MQTT, but there is something I can't figure out: setting outside temperature. 
I see from 
https://www.zigbee2mqtt.io/devices/SMT402.html
that it is supported in zigbee2mqtt, but I need to figure out what zigbee command is sent. 

I think the zigbee attribute Id is `0x4001`, although I could not confirm that since I could not find the mapping between `tz.stelpro_thermostat_outdoor_temperature` and that `0x4001` attribute despite grepping in 4 different repositories. At least, I know that writing to attribute 0x4001 does trigger a display of external temperature on my thermostat, but unfortunately, it does not display the value that I send (there must be some conversion needed). 

Question 1: where is this mapping between human readable names and zigbee attribute Ids done ?

I also need to figure out how is calculated the value that is sent with the zigbee command, and I could not find where that calculation is done in zigbee2mqtt. 

Question 2: where is that conversion done in zigbee2mqtt code ?

Thanks

**Comment by mboisson** [comment]
Ok, I found the answer to Question 1: 
https://github.com/Koenkk/zigbee-herdsman/blob/master/src/zcl/definition/cluster.ts#L2003

**Reply by mboisson** [reply]
And I guess the answer to question 2 is 
https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/converters/toZigbee.js#L4259

**Reply by Koenkk** [reply]
Both correct 😄

**Reply by mboisson** [reply]
So, the older Stelpro drivers for Hubitat did a bunch of things with the temperature: 
https://github.com/stelpro/maestro-thermostat/blob/master/devicetypes/stelpro/stelpro-maestro-thermostat.src/stelpro-maestro-thermostat.groovy#L332

i.e. 
```
    if (celsius < 0) {
        tempToSend = (celsius*100) + 65536
    }
    else {
    	tempToSend = (celsius*100)
    }
    tempToSendInString = zigbee.convertToHexString(tempToSend, 4)
``` 
and then, when writing the attribute, it did some hex and endianness magic on the hex string: 
https://github.com/stelpro/maestro-thermostat/blob/master/devicetypes/stelpro/stelpro-maestro-thermostat.src/stelpro-maestro-thermostat.groovy#L558
i.e. 
```
private String swapEndianHex(String hex) {
    reverseArray(hex.decodeHex()).encodeHex()
}

private byte[] reverseArray(byte[] array) {
    int i = 0;
    int j = array.length - 1;
    byte tmp;
    while (j > i) {
        tmp = array[j];
        array[j] = array[i];
        array[i] = tmp;
        j--;
        i++;
    }
    return array
}
``` 

Does any of this happens under the hood in zigbee2MQTT ? or is that simply no longer necessary for this model of thermostat ?

**Reply by Koenkk** [reply]
> Does any of this happens under the hood in zigbee2MQTT ?

This does not happen, code which sends this: https://github.com/Koenkk/zigbee-herdsman-converters/blob/1bfd652b91d1004505a2e8c684dfb16f77754461/converters/toZigbee.js#L4259

>  or is that simply no longer necessary for this model of thermostat ?

I don't have/know this thermostat so no clue.

**Reply by mboisson** [reply]
Thanks for the pointer. I should be able to figure things out with this info.
